osquery/tables/system/windows/windows_security_products.cpp

```diff
@@ -0,0 +1,203 @@
+  hr = PtrProductList->Initialize(provider);
+  if (FAILED(hr)) {
+    VLOG(1) << "Failed to initialize provider: " << hr;
+    return Status(hr);
+  }
+
+  // Get the number of security products of that type.
+  hr = PtrProductList->get_Count(&ProductCount);
+  if (FAILED(hr)) {
+    VLOG(1) << "Failed to get products count: " << hr;
+    return Status(hr);
+  }
+
+  // Loop over each product, querying the specific attributes.
+  for (LONG i = 0; i < ProductCount; i++) {
+    // Get the next security product
+    hr = PtrProductList->get_Item(i, &PtrProduct);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get product item: " << hr;
+      return Status(hr);
+    }
+    wsc_entry tmp;
+    tmp.provider = provider;
+
+    // Get the product name
+    hr = PtrProduct->get_ProductName(&PtrVal);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get product name: " << hr;
+      return Status(hr);
+    }
+    tmp.product_name = PtrVal;
+    SysFreeString(PtrVal);
+    PtrVal = nullptr;
+
+    // Get the product state
+    hr = PtrProduct->get_ProductState(&ProductState);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get product state: " << hr;
+      return Status(hr);
+    }
+    tmp.product_state = ProductState;
+
+    // Get the remediation path for the security product
+    hr = PtrProduct->get_RemediationPath(&PtrVal);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get remediation path: " << hr;
+      return Status(hr);
+    }
+    tmp.remediation_path = PtrVal;
+    SysFreeString(PtrVal);
+    PtrVal = nullptr;
+
+    // Get the signature status
+    hr = PtrProduct->get_SignatureStatus(&SignatureStatus);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get signature status: " << hr;
+      return Status(hr);
+    }
+    tmp.signature_status = SignatureStatus;
+
+    // Get the state timestamp
+    hr = PtrProduct->get_ProductStateTimestamp(&PtrVal);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get product state timestamp: " << hr;
+      return Status(hr);
+    }
+    tmp.product_state_timestamp = PtrVal;
+    SysFreeString(PtrVal);
+    PtrVal = nullptr;
+
+    PtrProduct->Release();
+    PtrProduct = nullptr;
+
+    out_list.push_back(tmp);
+  }
+
+  return Status(0);
+}
+
+void GetAllSecurityProducts(std::vector<wsc_entry>& out_list) {
+  if (!GetSecurityProducts(WSC_SECURITY_PROVIDER_FIREWALL, out_list).ok()) {
+    VLOG(1) << "Failed to get firewall products";
+  }
+  if (!GetSecurityProducts(WSC_SECURITY_PROVIDER_ANTIVIRUS, out_list).ok()) {
+    VLOG(1) << "Failed to get antivirus products";
+  }
+  if (!GetSecurityProducts(WSC_SECURITY_PROVIDER_ANTISPYWARE, out_list).ok()) {
+    VLOG(1) << "Failed to get antispyware products";
+  }
+}
+
+QueryData gen_wsp(QueryContext& context) {
+  QueryData results;
+  std::vector<wsc_entry> products;
+  GetAllSecurityProducts(products);
+  // Use this to convert std::wstring into std::string
+  auto& str_converter = std::wstring_convert<std::codecvt_utf8<wchar_t>>();
```

Comment:
Ouch, why do you save non-const reference to temporary object? Why is it ref?

Reply:
Happy to remove the reference here. This was probably just an oversight on my part.